lib/drivers/insulet/insuletDriver.js

```diff
@@ -1757,22 +1788,66 @@ module.exports = (config) => {
           }
 
           // no mounted drive with .ibf file, let's try connecting over MTP (Dash)
-          const WebMTP = require('webmtp');
-          const mtp = new WebMTP(deviceInfo.vendorId, deviceInfo.productId);
-
-          mtp.on('error', (err) => cb(err, null));
-          mtp.on('ready', async () => {
-            await mtp.openSession();
-            const handles = await mtp.getObjectHandles();
-            const objectHandle = Math.max(...handles);
-            const fileName = await mtp.getFileName(objectHandle);
-            bytes = await mtp.getFile(objectHandle, fileName);
-            buf = bytes.buffer;
-            await mtp.close();
-
+          const tempPath = (electron.app || electron.remote.app).getPath('temp');
+
+          if (process.platform !== 'win32') {
+            const mtp = require('node-mtp');
+            mtp.attach();
+            const list = mtp.getFileListing();
+            debug('Files:', list);
+            const tempFile = path.join(tempPath, 'temp.ibf');
+            const filtered = _.filter(list, (obj) => _.endsWith(obj.filename, '.ibf'));
+            const fileObj = _.maxBy(filtered, 'id');
+            debug('Retrieving', fileObj);
+            mtp.getFile(fileObj.id, tempFile);
+
+            const res = await fs.readFile(tempFile);
+            buf = res.buffer;
+            bytes = new Uint8Array(buf);
+            mtp.release();
+            await fs.unlink(tempFile);
             progress(100);
             return cb(null, data);
+          }
+          // using Powershell to read IBF file via MTP
+          debug('Storing .ibf file in', tempPath);
+
+          const Shell = require('node-powershell');
+          const ps = new Shell({
+            executionPolicy: 'Bypass',
+            noProfile: true,
           });
+
+          ps.addCommand('$o = New-Object -com Shell.Application');
+          ps.addCommand('$phoneObject = $o.NameSpace(17).items() | where { $_.name -eq "Android" }');
+          ps.addCommand('$source = $phoneObject.GetFolder.items() | where { $_.name -eq "Internal storage"}');
+          ps.addCommand('$sourceFolder = $o.NameSpace($source).self');
+          ps.addCommand('$item = $sourceFolder.GetFolder.Items() | Where-Object {$_.Type -eq "IBF File"} | select -Last 1');
+          ps.addCommand(`$destFolder = $o.NameSpace("${tempPath}").self`);
+          ps.addCommand('$destFolder.GetFolder.CopyHere($item)');
+          ps.addCommand('Write-Host $item.Name');
+
+          progress(50);
+          debug('Invoking PowerShell..');
+          const output = await ps.invoke();
+          debug('Copied', output);
+          progress(80);
+          ps.dispose();
+
+          if (output === '') {
```

Comment:
Could this check result in false positive if the PS script fails in any way? (e.g. prints out error message)

Reply:
Good question. For errors, `ps.invoke()` throws an error that is caught by the `try/catch` block, and I'm not aware of any error messages being printed to `output` when the script fails. This code is what is currently in production, so unless I have reason to believe that this is causing an issue, I'd rather not change it. I mean, if all goes well we can get rid of this code again in a month or two, right?

Reply:
👍🏻